File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int i=0, j=n-1;
        int[] ans={-1,-1};
        if(n==0) return ans;
        while(i <= j){
            int mid = i+(j-i)/2;
            if(nums[mid]==target){
                ans[0]= searchLeft(nums, target,0, mid);
                ans[1]=searchRight(nums, target, mid, n-1);
                return ans;
            }else if(nums[mid]>target){
                j=mid-1;
            }else{
                i=mid+1;
            }
        }
        return ans;
    }
    public int searchLeft(int[] nums, int target, int i, int j){
        while(i <= j){
            int mid=i+(j-i)/2;
            if(nums[mid] >= target){
                j=mid-1;
            }else{
                i=mid+1;
            }
        }
        return i;
    }
    public int searchRight(int[] nums, int target, int i, int j){
        while(i <= j){
            int mid=i+(j-i)/2;
            if(nums[mid] <= target){
                i=mid+1;
            }else{
                j=mid-1;
            }
        }
        return i-1;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int>ans={-1,-1};
        int low=0, high= nums.size();
        while(low<high){
            int mid = low+(high-low)/2;
            if(nums[mid]<=target){
                low=mid+1;
            }else{
                high=mid;
            }
        }
        if(low-1==-1 || nums[low-1]!=target){
            return ans;
        }
        int r=low-1;
        int l = findLow(nums, 0, r, nums[r]);
        ans[0]=l;
        ans[1]=r;
        return ans;
    }
    int findLow(vector<int>&nums, int low, int high, int t){
        while(low<high){
            int mid=low+(high-low)/2;
            if(nums[mid]<t){
                low=mid+1;
            }else{
                high=mid;
            }
        }
        return low;
    }
};